### ISSUE TITLE
Find substitute for /mnt directory

### ISSUE THREAD

**Issue by cbouchar** [beaker-project/restraint#74]
Currently restraint is using space beneath the /mnt directory structure.  These include the following directories:
    * scratchspace - Not currently used but will be soon.
    * testarea - contains taskout files plus current.log link 
    * tests - contains task distribution installs
We've discussed relocating this directory structure to another location since /mnt isn't appropriate and we can link /mnt to the new location.  What new directory do you have in mind?   I was looking at the following directory description 
https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard
and I came up with these suggestions:
    * /var/tmp/restraint - Even though this is temp, it persists between boots.
    * /var/lib/restraint - Currently we store our beaker provided config file, the
             environment file used by our commands, and avc_since which is used
             by our 10_avc_check, 10_avc_clear plugins.  It might be nice to keep
             everything in the same location.  An advantage to this directory is we know
             it works fine across our currently supported distros.
This change actually affects a number of other bugs so it needs to be brought up front.
This issue replaces BZ1813928.

**Comment by danrodrig** [comment]
> What new directory do you have in mind?

`/var/lib` State information. Persistent data modified by programs as they run.

`/var/lib/restraint/{scratchspace,testarea,tests}`

**Comment by StykMartin** [comment]
> > What new directory do you have in mind?
> 
> `/var/lib` State information. Persistent data modified by programs as they run.
> 
> `/var/lib/restraint/{scratchspace,testarea,tests}`

+1 on this. This is covered by macro `%{_sharedstatedir}`

**Comment by cbouchar** [comment]
I was leaning towards the latter one since it makes sense to centralize information.  I have yet to
review the Packaging documentation Martin mentioned  -->  https://docs.fedoraproject.org/en-US/packaging-guidelines/

**Comment by StykMartin** [comment]
Let's sum up the discussion.
- No upgrades between releases. It is expected that `restraintd` is always installed from scratch.
- Restraint should use `/var/lib` file structure by default.
- Restraint-RHTS should be running all workloads in the previous folder.

**Comment by cbouchar** [comment]
There are two caveats to this approach.
1. Performing an upgrade must be done by --force since /mnt has different attributes.  This is no longer an issue since Martin has chosen to eliminate support for rpm upgrade in this case.
2. If someone is performing 'relative' movements in directories, they will be broken.  Ex: if something created /mnt/mystuff and wrote code which performs ../mystuff, their task will be broken since mystuff doesn't exist in /var/lib/restraint.  I can't come up
with an exactly valid use case but this is a potential issue.
Option2 pull request 107 was created so you either work in /var/lib/restraint when using only restraint package or you work in /mnt directory when restraint-rhts package is installed.

**Comment by danrodrig** [comment]
> 1. Performing an upgrade must be done by --force since /mnt has different attributes.  This is no longer an issue since Martin has chosen to eliminate support for rpm upgrade in this case.

@StykMartin pointed out that the upgrade is not needed because restraint is always installed from scratch. There is a case where restraint is upgraded during a test. This happens while testing upgrades, for example, RHEL6 to RHEL7, but, the package is upgraded between distros, and the version is the same. From restraint 0.2.1 for RHEL6 to restraint 0.2.1 for RHEL7 for example. Therefore there is no issues with changing paths. Handling upgrades only makes sense for the restraint client.

All this was discussed during status meeting, and because of this we decided that we don't need to handle upgrades for restraint.

> 2. If someone is performing 'relative' movements in directories, they will be broken.  Ex: if something created /mnt/mystuff and wrote code which performs ../mystuff, their task will be broken since mystuff doesn't exist in /var/lib/restraint.  I can't come up
>    with an exactly valid use case but this is a potential issue.

I have doubts that this is a real issue. So, taking a look to how those directories are used,

The `/mnt/tests` path is used in restraint as `TASK_LOCATION` macro and it's used as the base path for the recipe, `recipe.base_path`.

The task path, `task.path`, contains the files for the task itself and it's used also as working directory during task execution.

In restraint, the tasks are installed or fetched into the `task.path` in the test system. There are two methods supported, RPM, for which the task is provided by installing an RPM, or URL, for which the task is fetched and extracted from an URL pointing to a tarball.

When the task is installed from RPM, the `task.path` is explicitly set in the job xml. For example,
```
<rpm name="distribution-distribution-command" path="/mnt/tests/distribution/command"/>
```
And the files are provided by the RPM itself,
```
# rpm -ql distribution-distribution-command
/mnt/tests
/mnt/tests/distribution
/mnt/tests/distribution/command
/mnt/tests/distribution/command/Makefile
/mnt/tests/distribution/command/PURPOSE
/mnt/tests/distribution/command/runtest.sh
/mnt/tests/distribution/command/testinfo.desc
```
When the task is fetched from URL, the `task.path` is the path of the url with `recipe.base_path` as base. For example, 
```
<fetch url="file:///usr/local/share/restraint-tasks/pointless.tar.gz"/>
```
The url path is, "/usr/local/share/restraint-tasks/pointless.tar.gz", and `recipe.base_path` being `/mnt/tests`, the `task.path` will be,
```
/mnt/tests/usr/local/share/restraint-tasks/pointless.tar.gz/
```
Notice that this is a directory,
```
$ ls -ld /mnt/tests/usr/local/share/restraint-tasks/pointless.tar.gz/
drwxr-xr-x. 2 root root 4096 Jul  1 07:38 /mnt/tests/usr/local/share/restraint-tasks/pointless.tar.gz/
```
Restraint itself will create the `task.path` directory and it will extract the task files into it.

Because of `task.path` structure, I don't think that escaping the symlink using a relative path is a real issue here. Doesn't make sense for a task to use a relative path for a parent. And if it does use it, the target is most likely to be another task installed or fetched by the recipe, therefore it should be under `TASK_LOCATION`, and no escaping should happen.

And going a little bit further, I don't think that the symlink `/mnt/tests` pointing to `TASK_LOCATION` needs to be provided at all. When RPM is used, the path is created by the RPM and the `task.path` is set from the job xml. And when URL is used, restraint will define and create the task path using `TASK_LOCATION` as base, extract everything there, and use it as working directory, regardless of what `TASK_LOCATION` path is.
Also note that restraint is setting `TASKPATH` environment variable from `task.path`.

There is one reference to `/mnt/tests` in `legacy/lib/rhts-make.include`. Not sure how it's used yet, so I don't know if this is an issue or not. But if it's an issue, the symlink should work.

The `/mnt/testarea` is used by restraint-rhts environment only. Unfortunately, this path was used also in the local watchdog plugins, but as commented in #105, this can be avoided easily. Therefore we can keep this path in the restraint-rhts package, and keep it out of restraint package and `/var/lib/restraint`. If something needs to be moved from restraint-rhts to restraint, that can be done later in a different issue. Meanwhile, restraint-rhts should be used. Priority is to remove `/mnt` from restraint, so this will make that happen sooner.

Is `/mnt/scratchspace` documented somewhere?

**Comment by cbouchar** [comment]
Search for 'CB -' for responses
> > 1. Performing an upgrade must be done by --force since /mnt has different attributes.  This is no longer an issue since Martin has chosen to eliminate support for rpm upgrade in this case.
> 
> @StykMartin pointed out that the upgrade is not needed because restraint is always installed from scratch. There is a case where restraint is upgraded during a test. This happens while testing upgrades, for example, RHEL6 to RHEL7, but, the package is upgraded between distros, and the version is the same. From restraint 0.2.1 for RHEL6 to restraint 0.2.1 for RHEL7 for example. Therefore there is no issues with changing paths. Handling upgrades only makes sense for the restraint client.
> 
> All this was discussed during status meeting, and because of this we decided that we don't need to handle upgrades for restraint.
> 
CB - Not sure why this needs further discussion since I already said Martin chose to eliminate upgrade support for restraint.

> > 1. If someone is performing 'relative' movements in directories, they will be broken.  Ex: if something created /mnt/mystuff and wrote code which performs ../mystuff, their task will be broken since mystuff doesn't exist in /var/lib/restraint.  I can't come up
> >    with an exactly valid use case but this is a potential issue.
> 
> I have doubts that this is a real issue. So, taking a look to how those directories are used,
> 
CB - No longer an issue as symlinks are not required as shown by my recent code changes.

> The `/mnt/tests` path is used in restraint as `TASK_LOCATION` macro and it's used as the base path for the recipe, `recipe.base_path`.

> 
> The task path, `task.path`, contains the files for the task itself and it's used also as working directory during task execution.
> 
> In restraint, the tasks are installed or fetched into the `task.path` in the test system. There are two methods supported, RPM, for which the task is provided by installing an RPM, or URL, for which the task is fetched and extracted from an URL pointing to a tarball.
> 
> When the task is installed from RPM, the `task.path` is explicitly set in the job xml. For example,
> 
> ```
> <rpm name="distribution-distribution-command" path="/mnt/tests/distribution/command"/>
> ```
> 
> And the files are provided by the RPM itself,
> 
> ```
> # rpm -ql distribution-distribution-command
> /mnt/tests
> /mnt/tests/distribution
> /mnt/tests/distribution/command
> /mnt/tests/distribution/command/Makefile
> /mnt/tests/distribution/command/PURPOSE
> /mnt/tests/distribution/command/runtest.sh
> /mnt/tests/distribution/command/testinfo.desc
> ```
> 
> When the task is fetched from URL, the `task.path` is the path of the url with `recipe.base_path` as base. For example,
> 
> ```
> <fetch url="file:///usr/local/share/restraint-tasks/pointless.tar.gz"/>
> ```
> 
> The url path is, "/usr/local/share/restraint-tasks/pointless.tar.gz", and `recipe.base_path` being `/mnt/tests`, the `task.path` will be,
> 
> ```
> /mnt/tests/usr/local/share/restraint-tasks/pointless.tar.gz/
> ```
> 
> Notice that this is a directory,
> 
> ```
> $ ls -ld /mnt/tests/usr/local/share/restraint-tasks/pointless.tar.gz/
> drwxr-xr-x. 2 root root 4096 Jul  1 07:38 /mnt/tests/usr/local/share/restraint-tasks/pointless.tar.gz/
> ```
> 
> Restraint itself will create the `task.path` directory and it will extract the task files into it.
> 
CB - This will be replaced by /var/lib/restraint/tests when the restraint-rhts package is NOT installed. This is reflected
in my recent code changes.
> Because of `task.path` structure, I don't think that escaping the symlink using a relative path is a real issue here. Doesn't make sense for a task to use a relative path for a parent. And if it does use it, the target is most likely to be another task installed or fetched by the recipe, therefore it should be under `TASK_LOCATION`, and no escaping should happen.
> 
CB - I don't believe a symlink is required at all.
> And going a little bit further, I don't think that the symlink `/mnt/tests` pointing to `TASK_LOCATION` needs to be provided at all. When RPM is used, the path is created by the RPM and the `task.path` is set from the job xml. And when URL is used, restraint will define and create the task path using `TASK_LOCATION` as base, extract everything there, and use it as working directory, regardless of what `TASK_LOCATION` path is.
> Also note that restraint is setting `TASKPATH` environment variable from `task.path`.
> 
CB - As you will see in my recent changes, there is no creation of /var/lib/restraint/tests as requested.  We'll leave this up to
the user to create a place for their report-results.

> There is one reference to `/mnt/tests` in `legacy/lib/rhts-make.include`. Not sure how it's used yet, so I don't know if this is an issue or not. But if it's an issue, the symlink should work.
CB - this reference is used in user Makefiles.  I'm leaving /mnt/tests unchanged for restraint-rhts package.  BTW here is
an example of Makefile which uses TEST_DIR defined by rhts-make.include.
https://gitlab.com/cbouchar/example-tasks/-/blob/master/setup_pass/Makefile
> 
> The `/mnt/testarea` is used by restraint-rhts environment only. Unfortunately, this path was used also in the local watchdog plugins, but as commented in #105, this can be avoided easily. Therefore we can keep this path in the restraint-rhts package, and keep it out of restraint package and `/var/lib/restraint`. If something needs to be moved from restraint-rhts to restraint, that can be done later in a different issue. Meanwhile, restraint-rhts should be used. Priority is to remove `/mnt` from restraint, so this will make that happen sooner.
CB - As you will see by my latest code changes, this is not included in /var/lib/restraint.
> 
> Is `/mnt/scratchspace` documented somewhere?
CB - I grepped around for this in rhts package and other places and have no idea why this was created.  I 
left rhts /mnt directories as is and will NOT carry this over to /var/lib/restraint.  This is reflected in my recent
code changes.

**Comment by danrodrig** [comment]
We agree on testarea and scratchspace, good. These two don't need to be moved to `/var/lib/restraint` and can be kept in `restraint-rhts`.

Now, `/mnt/tests`, I was thinking to not provide replacement at all in `restraint-rhts`, per what I wrote before. But, because that path in the metadata file, and the task that you have in GitLab, I'm not sure this is doable.

The code path in `restraint` itself to choose between `/mnt` or `/var/lib/restraint` has a major flaw, it will not work if restraint was installed from sources.

Alternative to that is what I already pointed out, to provide the `/mnt/tests` as symlink to `/var/lib/restraint/tests`. This should work fine, as escaping the symlink shouldn't be an issue, as I wrote before.

And another option that I thought when I saw the issue with the metadata file is to export `TASK_LOCATION` as an environment variable and use that variable to build `TEST_DIR`. Although `TEST_DIR` should be `TASK_PATH` or `TEST_PATH`, and they are already exported. I'm assuming here that whatever is importing that metadata file is within a task ran by restraint.

Between these two, the symlink is the easiest. The environment option, if doable, would help to make `TASK_LOCATION` configurable, but this can be done later.

**Comment by cbouchar** [comment]
I may be missing a piece of information and perhaps you can help.  Is it possible to install restraint by way of the restraint.spec file? I know you can perform yum-builddep restraint.spec file but that only installs dependencies.  Is there a way to
perform an install from source code via the restraint.spec file?  if not, rstraint-rhts will not be properly installed anyway since
it will be missing /mnt/testarea and a number of symbolic links provided by the spec file.  There will be tasks which will fail because of this.  So installing via source code may already have some major flaws.

**Comment by cbouchar** [comment]
I was under the impression the mission for the /mnt replacement change wasto replace the '/mnt' directories for non-rhts installations.  Some of thesuggestions don't meet this requirement so have we changed ourrequirements?
Here's what's on the table so far:

1.  Using link from /mnt to /var/lib/restraint does not achieve this but instead just adds redirection.  This change just adds noise to our code.
2. Use environment variables to replace /mnt.  This puts the change in the hands of the user and only the user that reads our    documentation. In this case, /mnt replacement is optional and in my opinion will less likely to occur.
3. Make /mnt replacement non-optional by finding a condition which indicates the rhts code is installed. If rhts is installed, use /mnt/tests; otherwise, use /var/lib/restraint/tests. Conditions can be 1) if any rhts package is installed or 2) one of the  rhts    scripts is in /usr/bin, or 3) test if /mnt/testarea is in place, or 4) ???other suggestions???.    In order to test for /mnt/testarea, the creation of /mnt/testarea would have to move to a legacy Makefile instead of restraint.spec file to meet Daniels requirements to successfully install from source code.

**Comment by cbouchar** [comment]
Fixed #107
